File: io/MultiStream.java
package common.io;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.HashMap;
import java.util.Map;

import common.CommonStatic;
import common.pack.Context.ErrType;

public class MultiStream {

	public static interface ByteStream {

		public void close() throws IOException;

		public void read(byte[] bs) throws IOException;

	}

	public static class TrueStream implements ByteStream {

		private final FileInputStream fis;

		public TrueStream(File f, int pos) throws IOException {
			fis = new FileInputStream(f);
			fis.skip(pos);
		}

		@Override
		public void close() throws IOException {
			fis.close();
		}

		@Override
		public void read(byte[] bs) throws IOException {
			fis.read(bs);
		}

	}

	private static interface RunExc {

		public void run() throws IOException;

	}

	private class SubStream implements ByteStream {

		private int pos;

		private SubStream(int pos) {
			this.pos = pos;
		}

		@Override
		public void close() throws IOException {
		}

		@Override
		public void read(byte[] bs) throws IOException {
			attempt(() -> readBytes(pos, bs));
			pos += bs.length;
		}

	}

	private static final Map<File, MultiStream> MAP = new HashMap<>();

	public static ByteStream getStream(File f, int pos, boolean useRAF) throws IOException {
		if (!useRAF) {
			return new TrueStream(f, pos);
		}
		MultiStream ms = MAP.get(f);
		if (ms == null)
			MAP.put(f, ms = new MultiStream(f));
		return ms.new SubStream(pos);
	}

	private final File file;
	private RandomAccessFile raf;
	private long poscache = -1;

	private MultiStream(File f) {
		file = f;
	}

	public void close() throws IOException {
		if (raf == null)
			return;
		poscache = -1;
		raf.close();
		raf = null;
	}

	private RandomAccessFile access() throws FileNotFoundException {
		if (raf != null)
			return raf;
		poscache = -1;
		raf = new RandomAccessFile(file, "r");
		return raf;
	}

	private void attempt(RunExc r) throws IOException {
		try {
			r.run();
			return;
		} catch (IOException e) {
			CommonStatic.ctx.printErr(ErrType.INFO, "failed to read, attempted again");
			close();
		}
		r.run();
		CommonStatic.ctx.printErr(ErrType.INFO, "attempt succeed");
	}

	private void readBytes(int pos, byte[] arr) throws IOException {
		access();
		if (poscache == -1)
			poscache = raf.getFilePointer();
		if (pos != poscache) {
			raf.seek(pos);
			poscache = pos;
		}
		raf.read(arr);
		poscache += arr.length;
	}

}

File: io/MultiStream.kt
package common.io

import common.CommonStatic
import common.pack.Context.ErrType
import java.io.*
import java.util.*

class MultiStream private constructor(private val file: File) {
    interface ByteStream {
        @Throws(IOException::class)
        fun close()

        @Throws(IOException::class)
        fun read(bs: ByteArray, i: Int, rlen: Int)
    }

    class TrueStream(f: File?, pos: Int) : common.io.MultiStream.ByteStream {
        private val fis: FileInputStream
        @Throws(IOException::class)
        override fun close() {
            fis.close()
        }

        @Throws(IOException::class)
        override fun read(bs: ByteArray, off: Int, len: Int) {
            fis.read(bs, off, len)
        }

        init {
            fis = FileInputStream(f)
            val skip = fis.skip(pos.toLong()) as Int
            if (skip != pos) throw IOException("failed to skip bytes")
        }
    }

    private interface RunExc {
        @Throws(IOException::class)
        fun run()
    }

    private inner class SubStream(private var pos: Int) : common.io.MultiStream.ByteStream {
        @Throws(IOException::class)
        override fun close() {
        }

        @Throws(IOException::class)
        override fun read(bs: ByteArray, off: Int, len: Int) {
            attempt { readBytes(pos, bs, off, len) }
            pos += bs.size
        }
    }

    private var raf: RandomAccessFile? = null
    private var poscache: Long = -1
    @Throws(IOException::class)
    fun close() {
        if (raf == null) return
        poscache = -1
        raf.close()
        raf = null
    }

    @Throws(FileNotFoundException::class)
    private fun access(): RandomAccessFile? {
        if (raf != null) return raf
        poscache = -1
        raf = RandomAccessFile(file, "r")
        return raf
    }

    @Throws(IOException::class)
    private fun attempt(r: common.io.MultiStream.RunExc) {
        try {
            r.run()
            return
        } catch (e: IOException) {
            CommonStatic.ctx.printErr(ErrType.INFO, "failed to read, attempted again")
            close()
        }
        r.run()
        CommonStatic.ctx.printErr(ErrType.INFO, "attempt succeed")
    }

    @Throws(IOException::class)
    private fun readBytes(pos: Int, arr: ByteArray, off: Int, len: Int) {
        access()
        if (poscache == -1L) poscache = raf.getFilePointer()
        if (pos.toLong() != poscache) {
            raf.seek(pos.toLong())
            poscache = pos.toLong()
        }
        raf.read(arr, off, len)
        poscache += len.toLong()
    }

    companion object {
        private val MAP: MutableMap<File, MultiStream> = HashMap()
        @Throws(IOException::class)
        fun getStream(f: File, pos: Int, useRAF: Boolean): common.io.MultiStream.ByteStream {
            if (!useRAF) {
                return TrueStream(f, pos)
            }
            var ms = MAP[f]
            if (ms == null) MAP[f] = MultiStream(f).also { ms = it }
            return ms!!.SubStream(pos)
        }
    }
}
